### PR TITLE
CRI-O: match version against openshift

### PR DESCRIPTION
**Pull request by runcom**
@mrunalp @kargakis @stevekuznetsov PTAL this will fix https://github.com/openshift/origin/pull/17546 by pulling the correct image for the CRI-O system container

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

### CODE REVIEW

**Comment by runcom** [comment]
Tag mapping lives here for now: https://hub.docker.com/r/runcom/cri-o/tags/

**Comment by stevekuznetsov** [comment]
Historically logic like this has a habit of breaking whenever something crazy happens with versioning. I think it may be better to have a switch-case and very explicit requirements?

**Comment by runcom** [comment]
You mean to not concatenate strings to build up the image string for the installer?  We can avoid that, but that requires us to update this repo every time a new release come up no? I think I'm missing something

**Comment by stevekuznetsov** [comment]
Yes, it would require it to be updated. If you think releases are very frequent then maybe it's a bad idea. If it's once per release of Origin that's usually once every couple of months and totally OK. I would rather heave a static list/mapping we have to update than logic that fails in some weird edge cases or ends up installing something unexpected. Switch/case can have a well defined `default`

File: sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml [hunk 1]
```diff
@@ -118,7 +119,7 @@ extensions:
                          --inventory sjb/inventory/ \
                          -e deployment_type=origin  \
                          -e openshift_use_crio=True   \
-                         -e openshift_crio_systemcontainer_image_override=docker.io/gscrivano/cri-o-centos \
+                         -e openshift_crio_systemcontainer_image_override="docker.io/runcom/cri-o-system-container:v${crio_tag}" \
```
Comment (ashcrow) [inline sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml]: Is ``crio_tag`` being set? In output that @mrunalp passed over to @sdodson and I the end result looked like:

```docker.io/runcom/cri-o-system-container:v```

Comment (giuseppe) [reply sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml]: true.  If this is the version currently used by the CI then `crio_tag` looks empty

Comment (ashcrow) [reply sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml]: This *seems* to be merged .... but not "merged". 😕

Comment (mrunalp) [reply sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml]: I just talked with @runcom He deployed it to test to see if it works. We need to debug why the tag isn't getting set.

Comment (runcom) [reply sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml]: yup, I'll debug it asap

Comment (ashcrow) [reply sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml]: @runcom were you able to find the issue?

Comment (runcom) [reply sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml]: I'm debugging here just right now, I'll update this once I find something

Comment (runcom) [reply sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml]: should be fixed now, I'll work on Steve's switch case for images

Comment (runcom) [reply sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml]: I've deployed this to the origin CI already so we can test if it fixes the empty crio_tag issue

**Comment by runcom** [comment]
@stevekuznetsov added the case statement for the crio image PTAL (deployed to the CI as well for the crio jobs)

File: sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml [hunk 2]
```diff
@@ -105,6 +105,21 @@ extensions:
       timeout: 7200
       repository: "aos-cd-jobs"
       script: |-
+        crio_tag="$( cat ./ORIGIN_PKG_VERSION | cut -d'-' -f2 | cut -d'.' -f1,2 )"
```
Comment (stevekuznetsov) [inline sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml]: I was also interested in _this_ being fragile. Can you use regex in the `case` statements?

Comment (runcom) [reply sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml]: You mean extracting something like "3.9" with a regexp from that file? I could do that yeah

Comment (stevekuznetsov) [reply sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml]: Right

**Comment by runcom** [comment]
@ashcrow so I've fixed the image tag but now we have another issue https://ci.openshift.redhat.com/jenkins/job/test_branch_origin_extended_conformance_crio/185/consoleFull#54402166758b6e51eb7608a5981914356

which is not related to CRI-O

**Comment by stevekuznetsov** [comment]
Are you starting the job manually? Use `/test`. The job expects many parameters to be set that were not provided in that run.

**Comment by runcom** [comment]
> Are you starting the job manually? Use /test. The job expects many parameters to be set that were not provided in that run.

oh, I see, I'll just /test crio then

**Comment by runcom** [comment]
So it's failing again https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_extended_conformance_crio/427/consoleFull#-190934373858b6e51eb7608a5981914356

but this time it tries to restart CRI-O but it hasn't been installed, has something changed in openshift-ansible that now forgets to "atomic install" CRI-O?

@giuseppe @ashcrow

**Comment by ashcrow** [comment]
@runcom not that I'm aware of ... but looking at the code it doesn't look correct. It literally installed node items but didn't install the system container 😕. 

I see that the ``docker`` role has been renamed ``container_runtime`` (which makes sense) ... I wonder if something slipped through the cracks.

**Comment by ashcrow** [comment]
I **think** [this](https://github.com/openshift/openshift-ansible/commit/5120f8e90c0178ac7f6d911159ceb278dd87b4c9#diff-0d092b0eb8108379abb4511b6bd1c47dL20) may be it but I'm not sure. The role used to require the ``docker`` role, but it no longer has a dependency.

@michaelgugino ^^ Thoughts?

**Comment by ashcrow** [comment]
Looks like it merged yesterday https://github.com/openshift/openshift-ansible/pull/6297

**Comment by runcom** [comment]
So that's another reason to test cri-o as well in openshift ansible ha

**Comment by michaelgugino** [comment]
Docker has been renamed to container_runtime to more accurately reflect its purpose.

This must now be called from prerequisites.yml before calling byo/config.yml

**Comment by ashcrow** [comment]
@michaelgugino thanks. Was there a reason this wasn't called already or is it a bug from refactoring?

**Comment by giuseppe** [comment]
I have some patches here, I will push them shortly.

Another thing, we should enable container-engine in the system-containers tests.

File: sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml (hunk 2 above)
Comment (runcom) [reply sjb/config/test_cases/test_branch_origin_extended_conformance_crio.yml]: fixed, now using grep

**Comment by sdodson** [comment]
> @michaelgugino thanks. Was there a reason this wasn't called already or is it a bug from refactoring?

It's intentional. So we need to ensure those vars are set when calling prerequisites too before this PR will produce the desired outcome.

**Comment by michaelgugino** [comment]
Any and all variables used with any and all of the plays must be used with prerequisites.yml.

In the case of container_runtime, that is all configured during prerequisites.yml.  prerequisites.yml configures all of the host-level settings for each container runtime.  In the case of system containers, this is when the runtime images are pulled and configured.

I highly recommend refactoring some of the tests to use a extra_vars file instead of passing most things via cli in the script.  This will help alleviate some of the issues in the future.

**Comment by mrunalp** [comment]
@michaelgugino Is that something which can be done in a follow-on PR?

**Comment by ashcrow** [comment]
@sdodson++

**Comment by mrunalp** [comment]
@sdodson Thanks!

**Comment by mrunalp** [comment]
@michaelgugino Does it look okay now with @sdodson commit?

**Comment by ashcrow** [comment]
I opened up https://github.com/runcom/cri-o-system-container/issues/1 to ensure that this is the image that is in @runcom's namespace. Pulling and looking at it I'm pretty sure it is.

**Pull request merged**